### ISSUE TITLE
RowToStructByName does not scan into embedded structures

### ISSUE THREAD

**Issue by SemenKuzmichevWork** [jackc/pgx#1717]
**Describe the bug**
I get "struct doesn't have corresponding row field id" error when I try to do RowToStructByName because I have another field in nested structure with `db:"id"` tag
**To Reproduce**
```go
type Subscription struct {
	ID          int       `db:"id"`
	TariffAlias string    `db:"tariff_alias"`
	Tariff
}

type Tariff struct {
	ID          int        `db:"id"`
	Alias       string     `db:"alias"`
}

pgx.CollectOneRow(rows, pgx.RowToStructByName[Subscription])
```

```sql
select * from subscription as s join tariff t on t.alias = s.tariff_alias;
```
**Expected behavior**
I get the completed structure Subscription
**Actual behavior**
struct doesn't have corresponding row field id

**Version**
 - Go: go version go1.21 darwin/amd64
 - pgx: github.com/jackc/pgx/v5.3.1

**Additional context**
The problem can be corrected by adding zeroing of the field name in the fieldPosByName function, as a mark that it has already found a match in the structure `fldDescs[i].Name = ""`

**Comment by jackc** [comment]
Perhaps we should have a better error message, but I don't think that your example *should* work. The `*StructByName*` helpers are supposed to work independently of column order. There is no way to safely scan multiple columns with the same name.

> The problem can be corrected by adding zeroing of the field name in the fieldPosByName function, as a mark that it has already found a match in the structure fldDescs[i].Name = ""

That might happen to work for your particular example query:

```sql
select * from subscription as s join tariff t on t.alias = s.tariff_alias;
```

But it could silently fail with data corruption if the column order was changed.

```sql
select * from tariff t join subscription as s on t.alias = s.tariff_alias;
```

**Comment by ivanpetrovs** [comment]
+1

**Comment by filipencopav** [comment]
+1

**Comment by filipencopav** [comment]
> Now, using **pgx.CollectOneRow(rows, pgx.RowToStructByName[User]),** I can join the user with the role

Is that so? I have an embedded struct into which I scan, and it doesn't see any embedded fields.

**Comment by jonathanmusto** [comment]
> > Now, using **pgx.CollectOneRow(rows, pgx.RowToStructByName[User]),** I can join the user with the role
> 
> Is that so? I have an embedded struct into which I scan, and it doesn't see any embedded fields.

Having the same problem here, nested struct fields aren't detected at all.

**Comment by filipencopav** [comment]
> Having the same problem here, nested struct fields aren't detected at all.

Check if the nested struct is a pointer. If it is, then it won't get scanned. If it's not a pointer, it will get scanned. Also, check that the fields are not private. If they are, it won't scan them.

```go
type Embedded struct {
    SomeField int32
    // this field will not be scanned whatsoever
    privateField float32
}

// this will work
type Parent struct {
    Embedded
}

// this will not work
type Parent2 struct {
    *Embedded
}
```

It would be cool if we were provided some ability to influence how fields are scanned. For example, we pass to a hypothetical `pgx.CustomRowToStruct(fn func(...))` a function that takes in a field and a row and returns a value/error. pgx takes care of all the scanning and iterating over the results boilerplate while we only say what to do with what field. Must also think about how to make it recursive into structs. Maybe the function would also need to take in the result.

**Comment by jonathanmusto** [comment]
> > Having the same problem here, nested struct fields aren't detected at all.
> 
> Check if the nested struct is a pointer. If it is, then it won't get scanned. If it's not a pointer, it will get scanned. Also, check that the fields are not private. If they are, it won't scan them.
> 
Yeah, I think that is the problem, my nested struct is a pointer. 

The problem I have is, these structs are all getting generated by the Buf CLI (https://buf.build/docs/) from my .proto file messages.  It seems to default to nesting as pointers with no way to override.

I'm currently playing around with scany, which looks like it might support pointers.

**Comment by filipencopav** [comment]
You can also write a custom scanner function, however it's complicated. I gave up, and instead made an intermediate struct to get all the scanned data, from which I then populate the real struct. Such are the fundamental limitations of [Place-Oriented-Programming (PLOP)](https://www.youtube.com/watch?v=-I-VpPMzG7c&t=390s).

**Comment by jflambert** [comment]
> Tell me, are there any changes on this topic? Is it possible to assemble this structure with one database query?
> 
> type Role struct {
> 	ID   int64   `db:"role_id"`
> 	Name string  `db:"role_name"`
> }
> 
> type Permission struct {
> 	ID     int64   `db:"permission_id"`
> 	Name   string  `db:"permission_name"`
> }
> 
> type User struct {
> 	ID       int64  `db:"user_id"`
> 	Email    string
> 	Password string
> 	Role
> 	Permissions []Permission
> }
> Now, using **pgx.CollectOneRow(rows, pgx.RowToStructByName[User]),** I can join the user with the role, but not the permissions. I think this is because permissions is a slice.
> 
> **Error:** There is no corresponding **permission_id** string field in the structure.

@eliofery can you show me what the `SELECT` looks like? I made sure to name all my db fields uniquely yet I'm having issues scanning into a simple nested structure (I don't have a slice)
